### PR TITLE
EKA-G18 | Adds on-init call back API

### PR DESCRIPTION
**Pull request by MounicaSagi**
* API gets patient link reference response as request body & puts it in redis cache against the original request id.

* adds tests for the API.

* New model has been introduced for patient link reference response.

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/link/link/Link.java  1
- src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5580492&bid=18292827)

File: src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+@JsonInclude(Include.NON_NULL)
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+@Data
+public class PatientLinkReferenceResult {
+    private UUID requestId;
+    private String timestamp;
+    private Link link;
+    private Error error;
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java:25]: Try to follow similar convention everywhere, In some places we have suffix response.

Check with people please.

Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java:25]: Angshuman have added /renamed error to RespError in other PR FYI.

File: src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    @NotNull
+    private GatewayResponse resp;
+
+    public boolean hasResponseId(){
+        if (resp == null) {
+            return false;
+        }
+        if (resp.getRequestId() == null || "".equals(resp.getRequestId())) {
+            return false;
+        }
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java:35]: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/StringUtils.html#StringUtils--

You can use this here.

Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java:35]: you can say 

return response != null && !StringUtils.isEmpty(resp.getRequestId());

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5580492&bid=18292827)

File: src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java (hunk 1 above)
Comment (MounicaSagi) [reply src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java]: Done.

File: src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java (hunk 2 above)
Comment (MounicaSagi) [reply src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java]: Done.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5580492&bid=18292827)

File: src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java (hunk 1 above)
Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java]: I think once we start deprecating and removing the older codes, we can reuse some of the names. Right now, we are struggling with names :)

File: src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java [hunk 3]
```diff
@@ -46,14 +46,16 @@ public Link link(WebClient.Builder builder,
                      CentralRegistry centralRegistry,
                      GatewayServiceProperties gatewayServiceProperties,
                      LinkServiceProperties serviceProperties,
-                     CacheAdapter<String, String> linkResults) {
+                     CacheAdapter<String, String> linkResults,
+                     CacheAdapter<String, String> patientLinkReferenceResults) {
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:50]: Do you need another CacheAdapter here? You can reuse the same linkResults - right?

File: src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java [hunk 4]
```diff
@@ -46,14 +46,16 @@ public Link link(WebClient.Builder builder,
...
         return new Link(
                 new LinkServiceClient(builder, centralRegistry, gatewayServiceProperties),
                 linkRepository,
                 centralRegistry,
                 serviceProperties,
-                linkResults);
-    }
+                linkResults,
+                patientLinkReferenceResults);
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:57]: I think there is no need to pass another CacheAdapter to the Link Service. Use the older linkResults. its ok.

File: src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java [hunk 5]
```diff
@@ -80,7 +82,7 @@ public Discovery discovery(DiscoveryRepository discoveryRepository,
     }
 
     @ConditionalOnProperty(value = "consentmanager.cacheMethod", havingValue = "guava", matchIfMissing = true)
-    @Bean({"discoveryResults", "linkResults"})
+    @Bean({"discoveryResults", "linkResults", "patientLinkReferenceResults"})
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:85]: not needed. use linkResults

Comment (MounicaSagi) [reply src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:85]: done

File: src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java (hunk 4 above)
Comment (MounicaSagi) [reply src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:57]: done

File: src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java (hunk 3 above)
Comment (MounicaSagi) [reply src/main/java/in/projecteka/consentmanager/link/LinkConfiguration.java:50]: done

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5580492&bid=18292827)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/model/PatientLinkReferenceResult.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5580492&bid=18292827)

**Pull request merged**